pkg/cache/queue/manager.go

```diff
@@ -462,15 +465,19 @@ func (m *Manager) AddOrUpdateWorkloadWithoutLock(log logr.Logger, w *kueue.Workl
 	}
 
 	qKey := queue.KeyFromWorkload(w)
-	m.deleteWorkloadFromQueuesIfReassigned(log, w, qKey)
+
+	assignedQueue, ok := m.workloadAssignedQueues[workload.Key(w)]
+	if ok && assignedQueue != qKey {
+		m.deleteAndForgetWorkload(log, w)
```

Comment:
Should we call here the public `DeleteAndForgetWorkload` method?

Reply:
The public variant has a lock. The private one does not. As such AddOrUpdateWorkloadWithoutLock requires the latter. I will rename the private name to mention no lock is set.